text/0135-compressed-blob-prefixes.md

```diff
@@ -0,0 +1,101 @@
+2. After the supermajority of production networks' nodes upgrades, one more change is released that adds `CBLOB_ZSTD_WASM_CODE` prefix instead of `CBLOB_ZSTD_WASM_OR_POV` when compiling and compressing Wasm parachain runtimes;
+3. Conservatively, wait until no more PVFs prefixed with `CBLOB_ZSTD_WASM_OR_POV` remain on-chain. That may take quite some time. Alternatively, create a migration that alters prefixes of existing blobs;
+4. Rename `CBLOB_ZSTD_WASM_OR_POV` to `CBLOB_ZSTD_POV`, as the PoVs remain the only entity compressed with this prefix by that time. Thus no collator protocol changes are needed.
+
+## Drawbacks
+
+Currently, the only requirement for a compressed blob prefix is not to coincide with Wasm magic bytes (as stated in code comments). Changes proposed here increase prefix collision risk, given that arbitrary data may be compressed in the future. However, it must be taken into account that:
+* Collision probability per arbitrary blob is ≈5,4×10⁻²⁰ for a single random 64-bit prefix (current situation) and ≈1,6×10⁻¹⁹ for the proposed set of three 64-bit prefixes (proposed situation), which is still low enough;
+* The current de facto protocol uses the current compression implementation to compress PoVs, which are arbitrary binary data, so the collision risk already exists and is not introduced by changes proposed here.
+
+## Testing, Security, and Privacy
+
+As the change increases granularity, it will positively affect both testing possibilities and security, allowing developers to check what's inside a given compressed blob precisely. Testing the change itself is trivial. Privacy is not affected by this change.
+
+## Performance, Ergonomics, and Compatibility
+
+### Performance
+
+The current implementation's performance is not affected by this change. Future implementations allowing for the execution of other-than-Wasm parachain runtimes will benefit from this change performance-wise.
+
+### Ergonomics
+
+The end-user ergonomics is not affected. The ergonomics for developers will benefit from this change as it enables exact checks and less guessing.
+
+### Compatibility
+
+The change is designed to be backward-compatible. 
+
+## Prior Art and References
+
+[SDK PR#6704](https://github.com/paritytech/polkadot-sdk/pull/6704) (WIP) introduces a mechanism similar to that described in this proposal and proves the necessity of such a change.
+
+## Unresolved Questions
+
+None
+
+## Future Directions and Related Material
+
+This proposal creates a foundation for two future work directions:
+* Proposing to introduce other-than-Wasm code executors, including PolkaVM, allowing parachain runtime authors to seamlessly change execution platform using the existing mechanism of runtime upgrades;
+* Proposing to use new compression methods that are faster/better/more secure than `zstd` as soon as they appear, streamlining the adoption of new methods and deprecation of the old ones.
```

Comment:
As far as I know it's somewhat unlikely that a significantly better general purpose compression codec than `zstd` will appear, at least probably in the next few decades, so this is somewhat of a weak motivation.

However, it would be feasible to have a *dedicated* (so not a general purpose) codec to compress certain types of blobs that could result in better compression and/or speed; for example, precomputing a zstd dictionary to compress `.wasm` blobs would most likely give us measurable improvements, but (even though it would still be compressed with zstd) it would require a mechanism (like what you propose here) to specify that such a precomputed dictionary should be used (because the dictionary wouldn't be transmitted, so it wouldn't be "normal" zstd compression).